### PR TITLE
Add a method (unused currently) to filter an input set

### PR DESCRIPTION
**Pull request by bobmcwhirter**
of purls to exlude any known-scanned recently enough.

### CODE REVIEW

File: collectorist/api/src/db/mod.rs [hunk 1]
```diff
@@ -183,6 +192,54 @@ impl Db {
...
+        let mut exclude_purls: Vec<String> = Vec::new();
+        let result = query.fetch_all(&self.pool).await?;
+
+        for row in result {
+            exclude_purls.push(row.get("purl"))
+        }
+
+        // We only remove a purl if we have definitively scanned it and have a
+        // record in our DB. Not-recently-enough or never-scanned will continue
+        // directly from the `input` vector to the output.
```
Comment (helio-frota) [inline collectorist/api/src/db/mod.rs:237]: thanks for the comments

File: collectorist/api/src/db/mod.rs [hunk 2]
```diff
@@ -329,4 +384,47 @@ mod test {
 
         Ok(())
     }
+
+    #[actix_web::test]
```
Comment (helio-frota) [inline collectorist/api/src/db/mod.rs:388]: I don't know this thing but if you added then I'm "ok"

File: collectorist/api/src/db/mod.rs [hunk 3]
```diff
@@ -183,6 +192,54 @@ impl Db {
         })
     }
 
+    async fn filter_purls_as_of(
+        &self,
+        collector_id: &str,
+        input: Vec<String>,
+        as_of: DateTime<Utc>,
+    ) -> Result<Vec<String>, anyhow::Error> {
+        // Per https://stackoverflow.com/questions/70029671/how-to-query-using-an-in-clause-and-a-vec-as-parameter-in-rust-sqlx-for-mysql
```
Comment (helio-frota) [inline collectorist/api/src/db/mod.rs:202]: cool I also do this kind of thing ( adding external URLs )

**Review by jcrossley3 — COMMENTED** [review]
Nice test and comments!

File: collectorist/api/src/db/mod.rs [hunk 4]
```diff
@@ -183,6 +192,54 @@ impl Db {
...
+        // We need to construct a substitution $3, $4, ... $n query for binding,
+        // as `sqlx` cannot expand a Vec<String> for use in an `IN` clause directly.
+        let mut in_params = String::new();
+
+        for i in 0..input.len() {
+            in_params.push_str(&format!("${}", i + 3));
+            if i < input.len() - 1 {
+                in_params.push(',');
+            }
+        }
```
[the remark this answers is not in the document]
Comment (helio-frota) [reply collectorist/api/src/db/mod.rs]: >  use enumerate, map and join here instead of len()

thanks , I'll save that tip

**Review by jcrossley3 — APPROVED** [review]

**Review by helio-frota — APPROVED** [review]

File: collectorist/api/src/db/mod.rs (hunk 3 above)
Comment (bobmcwhirter) [reply collectorist/api/src/db/mod.rs:202]: Nothing magic. Just external explanation as to what method I'm following, and why.

File: collectorist/api/src/db/mod.rs (hunk 2 above)
Comment (bobmcwhirter) [reply collectorist/api/src/db/mod.rs:388]: It's used to enable async tests. #[tokio::test] is also viable.

**Pull request merged**